Sprint 9 Project Submission #1

gjone124 opened the pull request.

gjone124 committed:
Sprint 9 Project Updated

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,106 @@
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then((res) => {
+      if (res.ok) {
+        return res.json();
+      }
+      return Promise.reject(`Error: ${res.status}`);
+    });
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special function `checkResponse` that will check any response from the server if it’s ok or not.


```js
function checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the function rather than the call. 

----------------------------------------------------------------

**COULD BE IMPROVED**

You can make a special function for fetching and checking responses **not to duplicate** it in every request:

```js

  function request(url, options) {
    return fetch(url, options).then(checkResponse)
  }

```

And now you can safely replace `fetch` with `request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,368 @@
+
+// Add Card (aka New Post) Modal Form Elements
+const addCardModal = document.querySelector("#add-card-modal");
+const addCardFormElement = addCardModal.querySelector(".modal__form");
```
**COULD BE IMPROVED**

We can get any form using their **unique** attribute `name` or `id` from `document.forms`

```js

const profileForm = document.forms["profile-form"];
const cardForm = document.forms["card-form"];

```

You need to add `name` or `id` first in `index.html` to achieve that.

More info here https://developer.mozilla.org/en-US/docs/Web/API/Document/forms

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,368 @@
+  event.preventDefault();
+
+  // changes text content to "Deleting..." (setButtonText is method from helpers.js)
+  const deleteButton = event.submitter;
```
Good job that you know about `event.submitter`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,368 @@
+function handleEditProfileFormSubmit(event) {
+  event.preventDefault();
+
+  // changes text content to "Saving..." (setButtonText is method from helpers.js)
+  const saveButton = event.submitter;
+  setButtonText(saveButton, true);
```
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,368 @@
+// Function To Display All Of The Cards From API Database
+function getInitialData() {
+  api
+    .getAppInfo()
+    .then(([cards, userInfo]) => {
```
Good job that you get the whole info with `Promise.all`

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Greg!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you **close** modals only in blocks `then`
* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Good job that you get the whole info with `Promise.all`

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* The `avatar` form texts should have `Poppins` font as well
* Please, add the same styles for the `avatar` modal
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special function `checkResponse` that will check any response from the server if it’s ok or not.


There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

gjone124 committed:
Sprint 9 Project Submission #2

gjone124 commented:
I implemented the 3 things that you said were essential to include / modify for the 2nd submission I just submitted. I added the checkResponse method (1), I made the avatar form text have Poppins font (2), and I made the avatar modal have the same styles as the other modals (3).

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
